src/lib.rs

```diff
@@ -989,22 +995,62 @@ pub async fn up(docker: &Docker, config: &Config) -> Result<(), Error> {
         debug!("Waiting until port {} is available", mapping.host);
         tokio::time::sleep(std::time::Duration::new(5, 0)).await;
     }
+    debug!("Port is available, proceeding");
+
+    // Sometimes, Synapse fails to launch, either because it cannot open the
+    // port (possibly Unix isn't actually closing it fast enough) or because
+    // it receives a SIGTERM for reasons so far unknown.
+    //
+    // The best workaround we have atm is to retry Synapse launch until it
+    // seems to have worked.
+    let mut synapse_is_responsive = false;
+    'try_to_launch_synapse: for i in 0..MAX_ATTEMPTS_TO_START_SYNAPSE {
+        // It's now time to run Synapse.
+        start_synapse_container(
+            docker,
+            config,
+            &synapse_data_directory,
+            &run_container_name,
+            vec!["/start.py".to_string()],
+            true,
+        )
+        .await
+        .context("Failed to start Synapse")?;
 
-    // It's now time to run Synapse.
-    start_synapse_container(
-        docker,
-        config,
-        &synapse_data_directory,
-        &run_container_name,
-        vec!["/start.py".to_string()],
-        true,
-    )
-    .await
-    .context("Failed to start Synapse")?;
+        debug!("Synapse is now launched (attempt {})", i);
+
+        for j in 0..MAX_ATTEMPTS_TO_WAIT_FOR_SYNAPSE {
+            if is_synapse_ready(config).await? {
+                synapse_is_responsive = true;
+                break 'try_to_launch_synapse;
+            }
+            tokio::time::sleep(std::time::Duration::from_millis(
+                SLEEP_WHILE_SYNAPSE_STARTS_MS,
+            ))
+            .await;
+            if !docker.is_container_running(&run_container_name).await? {
+                // Synapse startup didn't work out for reasons unknown.
+                continue 'try_to_launch_synapse;
```

Comment:
So we expect that synapse can be "not ready" but the container can be running, and this is when we "wait for Synapse"? But if the container isn't running we just skip to relaunching it? 

I mean using labels is probably ok here but it would be better if they could be included in a loop condition, I think you have probably thought about that though. if `for` could be combined with `while` we could probably avoid it but i don't think that's a thing

Reply:
> So we expect that synapse can be "not ready" but the container can be running, and this is when we "wait for Synapse"? But if the container isn't running we just skip to relaunching it?

Exactly.

> I mean using labels is probably ok here but it would be better if they could be included in a loop condition, I think you have probably thought about that though. 

I'llsee if I can find a way to make this more readable.